src/netconf_acm.c

```diff
@@ -0,0 +1,1862 @@
+ */
+static sr_error_info_t *
+sr_nacm_allowed_tree(const struct lysc_node *root, const char *user, int* allowed)
+{
+    int user_found;
+    sr_error_info_t *err_info = NULL;
+    uid_t user_uid;
+
+    /* 1) NACM is off */
+    if (!nacm.enabled) {
+        *allowed = 1;
+        return err_info;
+    }
+
+    /* 2) recovery session allowed */
+    err_info = sr_nacm_getpwnam(user, &user_uid, NULL, &user_found);
+    if (err_info) {
+        return err_info;
+    }
+
+    if (user_found && user_uid == SR_SU_UID) {
+        *allowed = 1;
+        return err_info;
+    }
+
+    /* 3) <close-session> and notifications <replayComplete>, <notificationComplete> always allowed */
+    if ((root->nodetype == LYS_RPC) && !strcmp(root->name, "close-session") &&
+            !strcmp(root->module->name, "ietf-netconf")) {
+        *allowed = 1;
+        return err_info;
+    } else if ((root->nodetype == LYS_NOTIF) && !strcmp(root->module->name, "nc-notifications")) {
+        *allowed = 1;
+        return err_info;
+    }
+
+    /* 4) <get>, <get-config>, and <get-data> not checked for execute permission - RFC 8341 section 3.2.4
+     * (assume it is the same for <get-data>) */
+    if ((root->nodetype == LYS_RPC) && (((!strcmp(root->name, "get") || !strcmp(root->name, "get-config")) &&
+            !strcmp(root->module->name, "ietf-netconf")) || (!strcmp(root->name, "get-data") &&
+            !strcmp(root->module->name, "ietf-netconf-nmda")))) {
+        *allowed = 1;
+        return err_info;
+    }
+
+    *allowed = 0;
+    return err_info;
+}
+
+/**
+ * @brief Collect all NACM groups for a user. If enabled, even system ones.
+ *
+ * @param[in] user User to collect groups for.
+ * @param[out] groups Sorted array of collected groups.
+ * @param[out] group_count Number of @p groups.
+ * @return err_info, NULL on success.
+ */
+static sr_error_info_t *
+sr_nacm_collect_groups(const char *user, char ***groups, uint32_t *group_count)
+{
+    sr_error_info_t *err_info = NULL;
+    struct group grp, *grp_p;
+    gid_t user_gid;
+    char *buf = NULL;
+    gid_t *gids = NULL;
+    ssize_t buflen;
+    uint32_t i, j;
+    int found;
+    int gid_count = 0, ret;
+
+    *groups = NULL;
+    *group_count = 0;
+
+    /* collect NACM groups */
+    for (i = 0; i < nacm.group_count; ++i) {
+        for (j = 0; j < nacm.groups[i].user_count; ++j) {
+            if (!strcmp(nacm.groups[i].users[j], user)) {
+                if (sr_nacm_strarr_sort_add((const char **)&nacm.groups[i].name, sizeof **groups, 0, groups, group_count)) {
+                    goto cleanup;
+                }
+            }
+        }
+    }
+
+    /* collect system groups */
+    if (nacm.enable_external_groups) {
+        err_info = sr_nacm_getpwnam(user, NULL, &user_gid, &found);
+        if (err_info) {
+            goto cleanup;
+        }
+
+        if (!found) {
+            /* no user, no more groups */
+            sr_errinfo_free(&err_info);
+            err_info = NULL;
+        }
```

Comment:
Do not understand the purpose of this code. There should probably just be `goto cleanup;` because there is nothing else to do.